src/modelBasedFixers/apex-pmd-violations-fixer.ts

```diff
@@ -0,0 +1,164 @@
+        : String(error);
+        console.error(errorMessage, error);
+        void vscode.window.showErrorMessage(`${errorMessage} ${detailedMessage}`);
+
+        // We still need to provide the command arguments to avoid unintended errors being surfaced to the user
+        const [document] = codeAction.command.arguments as [vscode.TextDocument];
+        codeAction.command.arguments = [document.getText(), document.uri];
+        return codeAction;
+      }
+    })();
+  }
+
+  public removeCodeMarkdowns(codeSnippet: string) {
+    if (codeSnippet.startsWith('```apex')) {
+      codeSnippet = codeSnippet.trimEnd();
+      codeSnippet = codeSnippet.slice(7, codeSnippet.length - 3);
+    }
+    return codeSnippet;
+  }
+
+  public generatePrompt(document:vscode.TextDocument, diagnostic: vscode.Diagnostic): string {
+    const promptFormatter = new PromptFormatter(PromptConstants.SYSTEM_PROMPT.concat(PromptConstants.BASE_USER_PROMPT));
+    return promptFormatter.substitute(
+        this.extractCodeFromFile(document.getText(), diagnostic.range.start.line + 1, diagnostic.range.end.line + 1),
+        diagnostic.message, PromptConstants.APEX_CRUD_VIOLATION_PROMPT);
+  }
+
+  public extractCodeFromFile(
+    fileContent: string,
+    startLine: number,
+    endLine: number
+): string | null {
+    // Split the file content into lines for processing
+    const lines = fileContent.split('\n');
+
+    // Validate the line numbers
+    if (startLine < 1 || endLine > lines.length || startLine > endLine) {
+        throw new Error('Invalid start or end line number.');
+    }
+
+    // Extract the lines between startLine and endLine (inclusive)
+    const extractedLines = lines.slice(startLine - 1, endLine);
+    return extractedLines.join('\n');
+  }
+
+
+  public replaceCodeInFile(
+    fileContent: string,
+    replaceCode: string,
+    startLine: number,
+    endLine: number
+): string {
+    // Split the file content into an array of lines
+    const lines = fileContent.split('\n');
+
+    if (startLine < 1 && endLine >= lines.length - 1) {
+      // This means the whole file content is replaced.
+      // This happens for some fixes like the OverrideBothEqualsAndHashcode violation fix where the model returns the entire class as fix.
+      // This should not happen eventually and only the fixed lines will be returned by the model.
+      return replaceCode;
+    }
+
+    // Ensure startLine and endLine are within valid ranges
+    if (startLine < 1 || endLine > lines.length || startLine > endLine) {
+        throw new Error('Invalid startLine or endLine values.');
+    }
+
+    // Determine the leading spaces of the first line being replaced
+    const leadingSpaces = lines[startLine - 1].match(/^\s*/)?.[0] || '';
+
+    // Add the leading spaces to only the first line of the replaceCode
+    const [firstLine, ...remainingLines] = replaceCode.split('\n');
+    const indentedReplaceCode = [leadingSpaces + firstLine, ...remainingLines].join('\n');
```

Comment:
Why do we only indent the first line? Do the rest of the lines get auto-indented?

Reply:
Yes, the rest of the lines get auto-indented.